network-area-diagram/src/main/java/com/powsybl/nad/svg/DiagramMetadata.java

```diff
@@ -0,0 +1,206 @@
+        String diagramId = reader.getAttributeValue(null, DIAGRAM_ID_ATTRIBUTE);
+        String equipmentId = reader.getAttributeValue(null, EQUIPMENT_ID_ATTRIBUTE);
+        return new DeserializedIdentifiable(diagramId, equipmentId);
+    }
+
+    private static Point parsePoint(XMLStreamReader reader) {
+        double x = parseDouble(reader.getAttributeValue(null, POSITION_X_ATTRIBUTE));
+        double y = parseDouble(reader.getAttributeValue(null, POSITION_Y_ATTRIBUTE));
+        return new Point(x, y);
+    }
+
+    private static double parseDouble(String s) {
+        if (s == null || s.isEmpty()) {
+            return Double.NaN;
+        }
+        return Double.parseDouble(s);
+    }
+
+    interface MetadataItemWriter<I extends IdentifiableMetadata> {
+        // To avoid checked exceptions while writing specific data for metadata elements
+        void write(I item, XMLStreamWriter writer) throws XMLStreamException;
+    }
+
+    public void writeXml(XMLStreamWriter writer) throws XMLStreamException {
+        writeXml(UnaryOperator.identity(), writer);
+    }
+
+    public void writeXml(UnaryOperator<String> diagramIdToSvgId, XMLStreamWriter writer) throws XMLStreamException {
+        writer.writeStartElement(METADATA_PREFIX, METADATA_DIAGRAM_ELEMENT_NAME, METADATA_NAMESPACE_URI);
+        writer.writeNamespace(METADATA_PREFIX, METADATA_NAMESPACE_URI);
+        writeCollection(METADATA_BUS_NODES_ELEMENT_NAME, METADATA_BUS_NODE_ELEMENT_NAME, busNodesMetadata, DiagramMetadata::writeNodeMetadata, diagramIdToSvgId, writer);
+        writeCollection(METADATA_NODES_ELEMENT_NAME, METADATA_NODE_ELEMENT_NAME, nodesMetadata, DiagramMetadata::writeNodeMetadata, diagramIdToSvgId, writer);
+        writeCollection(METADATA_EDGES_ELEMENT_NAME, METADATA_EDGE_ELEMENT_NAME, edgesMetadata, DiagramMetadata::writeEdgeMetadata, diagramIdToSvgId, writer);
+        writer.writeEndElement();
+    }
+
+    private static <I extends IdentifiableMetadata> void writeCollection(
+            String collectionElementName,
+            String itemElementName,
+            Collection<I> items,
+            MetadataItemWriter<I> elemWriter,
+            UnaryOperator<String> diagramIdToSvgId,
+            XMLStreamWriter writer) throws XMLStreamException {
+        if (items.isEmpty()) {
+            writer.writeEmptyElement(METADATA_PREFIX, collectionElementName, METADATA_NAMESPACE_URI);
+        } else {
+            writer.writeStartElement(METADATA_PREFIX, collectionElementName, METADATA_NAMESPACE_URI);
+            for (I item : items) {
+                writer.writeEmptyElement(METADATA_PREFIX, itemElementName, METADATA_NAMESPACE_URI);
+                writer.writeAttribute(DIAGRAM_ID_ATTRIBUTE, diagramIdToSvgId.apply(item.getIdentifiable().getDiagramId()));
+                writer.writeAttribute(EQUIPMENT_ID_ATTRIBUTE, item.getIdentifiable().getEquipmentId());
+                elemWriter.write(item, writer);
```

Comment:
You don't need a new interface `MetadataItemWriter`, you could use directly `IdentifiableMetadata`:
```java
item.write(writer);
```
Then you need to add `getElementName` to have a default implementation in `IdentifiableMetadata`, and a `svgId` attribute in it also

Reply:
refactored the metadata items so they know how to write themselves to xml. Also refactored the reads. The diagram metadata only knows about serialization/deserialization of the collections it owns.